common/src/Microsoft.Azure.IIoT.AspNetCore/src/Auth/Clients/MsalUserTokenClient.cs

```diff
@@ -0,0 +1,305 @@
+            _schemes = schemes ?? throw new ArgumentNullException(nameof(schemes));
+            _ctx = ctx ?? throw new ArgumentNullException(nameof(ctx));
+            _cache = cache ?? throw new ArgumentNullException(nameof(cache));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _config = config ?? throw new ArgumentNullException(nameof(config));
+        }
+
+        /// <inheritdoc/>
+        public bool Supports(string resource) {
+            return _config.Query(resource, Provider).Any();
+        }
+
+        /// <inheritdoc/>
+        public async Task<TokenResultModel> GetTokenForAsync(string resource,
+            IEnumerable<string> scopes) {
+            if (_ctx.HttpContext?.User == null) {
+                return null;
+            }
+            var schemes = await _schemes.GetAllSchemesAsync();
+            if (!schemes.Any(s => s.Name == Provider)) {
+                return null;
+            }
+            foreach (var config in _config.Query(resource, Provider)) {
+                var decorator = CreateConfidentialClientApplication(_ctx.HttpContext.User,
+                    config, CreateRedirectUrl());
+                try {
+                    var result = await AcquireTokenSilentAsync(decorator.Client,
+                        _ctx.HttpContext.User, GetScopes(config, scopes), config.TenantId);
+                    if (result != null) {
+                        _logger.Debug(
+                            "Successfully acquired token {resource} with {config}.",
+                            resource, config.GetName());
+                        return result.ToTokenResult();
+                    }
+                }
+                catch (MsalUiRequiredException e) {
+                    var validatedToken = (JwtSecurityToken)_ctx.HttpContext?.Items["pass_through"];
+                    if (validatedToken != null) {
+                        // to get a token for a web api on behalf of the user inside a web api.
+                        // In the case the token is a JWE (encrypted token), we use the decrypted token.
+                        var accessToken = validatedToken.InnerToken == null ?
+                            validatedToken.RawData : validatedToken.InnerToken.RawData;
+
+                        var result = await decorator.Client.AcquireTokenOnBehalfOf(
+                            GetScopes(config, scopes), new UserAssertion(accessToken))
+                            .ExecuteAsync();
+                        _logger.Information(
+                            "Successfully acquired on behalf token for {resource} with {config}.",
+                                resource, config.GetName());
+                        return result.ToTokenResult();
+                    }
+                    else {
+                        _logger.Debug(e, "Failed to get token for {resource} with {config}.",
```

Comment:
>Debug [](start = 32, length = 5)

shouldn't this be a warning or info?

Reply:
Want to avoid too much spewing here as some of this is by design.  I was thinking to pass a "context" through all token clients where each can add some diagnostics, and if no token can be provided the entire diagnostics chain is dumped out.   I will add a workitem.

Reply:
Actually added it so that all exceptions are pulled up as aggregate exception and then logged when no token could be retrieved.